source/docs/software/basic-programming/functions-as-data.rst

```diff
@@ -0,0 +1,118 @@
+   public void resetEncoders()
+
+If the function signature does not match this, Java will not be able to interpret the method reference as a ``Runnable`` and the code will not compile.
+
+Lambda Expressions in Java
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If we do not already have a named function that does what we want, we can define a function "inline" - that means, right inside of the call to ``runOnce``!  We do this by writing our function with a special syntax that uses an "arrow" symbol to link the argument list to the function body:
+
+.. code-block:: java
+
+   // Create an InstantCommand that runs the drive forward at half speed
+   Command driveHalfSpeed = runOnce(() -> { drivetrain.arcadeDrive(0.5, 0.0); }, drivetrain);
+
+Java calls ``() -> { drivetrain.arcadeDrive(0.5, 0.0); }`` a "lambda expression"; it may be less-confusingly called an "arrow function", "inline function", or "anonymous function" (because it has no name).  While this may look a bit funky, it is just another way of writing a function - the parentheses before the arrow are the function's argument list, and the code contained in the brackets is the function body.  The "lambda expression" here represents a function that calls ``drivetrain.arcadeDrive`` with a specific set of parameters - note again that this does not *call* the function, but merely defines it and passes it to the ``Command`` to be run later when the ``Command`` is scheduled.
+
+Note that our inline function still has to be a ``Runnable`` - notice that it takes no arguments and has no return statement.  If it did not match the ``Runnable`` contract, our code would fail to compile.
+
+Capturing State in Java Lambda Expressions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the above example, our function body references an object that lives outside of the function itself (namely, the ``drivetrain`` object).  This is called a "capture" of a variable from the surrounding code (which is sometimes called the "outer scope" or "enclosing scope").  Usually the captured variables are either local variables from the enclosing method body in which the lambda expression is defined, or else fields of an enclosing class definition in which that method is defined.
+
+In Java capturing state is a fairly safe thing to do in general, with one major caveat: we can only capture state that is "effectively final".  That means it is only legal to capture a variable from the enclosing scope if that variable is never reassigned after initialization.  Note that this does not mean that the captured state cannot change: Remember that Java objects are references, so the object that the reference *points to* may change after capture - but the reference itself cannot be made to point to another object.
+
+This means we can only capture primitive types (like ``int``, ``double``, and ``boolean``) if they're constants.  If we want to capture a state variable that can change, it *must be wrapped in a mutable object*.
```

Comment:
Will violating this lead to compilation failure or the lambda not picking up subsequent changes to the value? I think VS Code warns about it sometimes but I don't recall if it fails compilation. Probably worth including a bit on how someone can know they messed this up.

Reply:
It'll lead to compilation failure.